.goreleaser.yml

```diff
@@ -0,0 +1,42 @@
+# This is an example goreleaser.yaml file with some sane defaults.
+# Make sure to check the documentation at http://goreleaser.com
+before:
+  hooks:
+    - go mod tidy
+builds:
+- main: ./cmd/gantry/gantry.go
+  env:
+    - CGO_ENABLED=0
+  goos:
+    - linux
+  goarch:
+    - amd64
+    - arm
+    - arm64
+    - ppc64le
+    - s390x
+  goarm:
+    - 6
+    - 7
+  ldflags:
+    - -X "github.com/ad-freiburg/gantry.Version={{ .Env.GANTRY_VERSION }}"
+archives:
+- name_template: "{{ .ProjectName }}_{{ .Os }}_{{ .Arch }}{{ if .Arm }}v{{ .Arm }}{{ end }}"
```

Comment:
I think this doesn't match the `uname -m` used in our install instructions. On my Pi 4 I get `armv7l` so it's missing the `l`?

Reply:
Hmm, to be honest I never tried that just used the defaults and hoped they work. Currently non of my (mostly older) pi boot...

I'll have to look into this ... don't really know why there is an `l` in the `uname` and such things

Reply:
**tl;dr it's complicated and more testing needed**

If I've read it correctly `armv7` describes the general architecture which can be extended by adding suffixes like `l` standing for `Little Endian`.
The debian version [0] is named `armhf` in it, `hf` standing for `hard float`. The suffix for the architecture would be `h` [1] thus `armv7h`.
If [2] is correct, RP3 (and i think RP4 too) is an `armv8` running in 32bit mode and calling itself `armv7`.

The tests run by travis run on `arm64`/`armv8` [3] thus they don't say anything about running on a pi (at least in 32bit mode).

goreleaser just uses `GOARCH` and `GOARM` [4], these do not contain such extensions as `l` or `h`, but `GOARM=6` implies some floating-point support from the processor, looking at the supported architectures [5] `armv8` is it's own arch.

**Stuff to test:**
Can you please send me the output of `uname` with the following options?
```
  -m, --machine            print the machine hardware name
  -p, --processor          print the processor type (non-portable)
  -i, --hardware-platform  print the hardware platform (non-portable)
```
Maybe `-m` can be replaced with another one, for my laptop I get these results:
```
% uname -a
Linux tender 5.3.11-300.fc31.x86_64 #1 SMP Tue Nov 12 19:08:07 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux
% uname -m
x86_64
% uname -p
x86_64
% uname -i
x86_64
```

[0] https://wiki.debian.org/ArmHardFloatPort
[1] https://stackoverflow.com/a/29167402
[2] https://www.raspberrypi.org/forums/viewtopic.php?t=140572
[3] https://blog.travis-ci.com/2019-10-07-multi-cpu-architecture-support
[4] https://golang.org/doc/install/source#environment
[5] https://github.com/golang/go/wiki/GoArm#supported-architectures

Reply:
I think most architectures are a lot less complicated than ARM as they (and s390x especially) take binary compatibility much more seriously. I'll test the `-p` and `-i` options at work tomorrow but `-m` should be `s390x` on all systems supported by Linux (at least back to the z900 from 2000). As for POWER there should only be `ppc64le` and `ppc64be` but the latter is rare when running Linux. I think for ARM64 (ARM wants us to call it AArch64 but Apple made sure no one will ever use that terrible name) it should just be `armv8` too. So that leaves the mess for 32 bit ARM, do you know if Raspbian uses `armv7l` too?

Reply:
Not looking good that good as only `-m` produces usable output on my Pis.

The following commands where performed on a freshly written SDCard using [2019-09-26-raspbian-buster-lite.zip](downloads.raspberrypi.org/raspbian_lite/images/raspbian_lite-2019-09-30/2019-09-26-raspbian-buster-lite.zip)

Raspberry Pi Model B Plus Rev 1.2
```
pi@raspberrypi:~ $ uname -a
Linux raspberrypi 4.19.75+ #1270 Tue Sep 24 18:38:54 BST 2019 armv6l GNU/Linux
pi@raspberrypi:~ $ uname -m
armv6l
pi@raspberrypi:~ $ uname -p
unknown
pi@raspberrypi:~ $ uname -i
unknown
```

Raspberry Pi 2 Model B Rev 1.1
```
pi@raspberrypi:~ $ uname -a
Linux raspberrypi 4.19.75-v7+ #1270 SMP Tue Sep 24 18:45:11 BST 2019 armv7l GNU/Linux
pi@raspberrypi:~ $ uname -m
armv7l
pi@raspberrypi:~ $ uname -p
unknown
pi@raspberrypi:~ $ uname -i
unknown
```

Raspberry Pi 3 Model B Rev 1.2
```
pi@raspberrypi:~ $ uname -a
Linux raspberrypi 4.19.75-v7+ #1270 SMP Tue Sep 24 18:45:11 BST 2019 armv7l GNU/Linux
pi@raspberrypi:~ $ uname -m
armv7l
pi@raspberrypi:~ $ uname -p
unknown
pi@raspberrypi:~ $ uname -i
unknown
```
I've then tried retrieved information using ArchLinuxARM following the instructions on the wiki page for [Raspberry Pi](https://archlinuxarm.org/platforms/armv6/raspberry-pi). But my 3B+ does not boot, I only see the color screen. I'd have liked to report if this distro reported different values.

Reply:
So on IBM Z `-m`, `-p`, and `-i` are all `s390x` tested on z14 and z15 (but I'm sure z13 matches that as well).

On my RaspberryPi 4 running Arch Linux ARM I get:
```
niklas@mimas ~ % uname -a
Linux mimas 4.19.85-1-ARCH #1 SMP PREEMPT Sat Nov 23 22:50:37 UTC 2019 armv7l GNU/Linux
niklas@mimas ~ % uname -m
armv7l
niklas@mimas ~ % uname -i
unknown
niklas@mimas ~ % uname -p
unknown
```

I also got an Odroid C2 ARM64 system that shows the following:
```
edgy@edgy-c2-40:~$ uname -a
Linux edgy-c2-40 4.17.3-odroidc2 #3 SMP PREEMPT Thu Jun 28 10:02:07 UTC 2018 aarch64 GNU/Linux
edgy@edgy-c2-40:~$ uname -m
aarch64
edgy@edgy-c2-40:~$ uname -i
unknown
edgy@edgy-c2-40:~$ uname -p
unknown
```